### PR TITLE
Pass extended key usage parameters when importing CertificateSigningRequestParams 

### PR DESCRIPTION
**Pull request by uglyoldbob**
This pull requests adds parsing of most of the extended key usage parameters, except for Other.

### CODE REVIEW

File: rcgen/src/csr.rs
Comment (djc) [inline rcgen/src/csr.rs:119]: This is pretty verbose. Suggest adding a private method `fn insert_extended_key_usage(&mut self, eku: ExtendedKeyUsagePurpose)` to `CertificateParams` (at the bottom of the `impl`) that does the `contains()`/`push()` dance.

**Review by djc — COMMENTED** [review]
Would be good to support this, thanks for the PR! Suggesting a more concise way of doing it.

**Review by est31 — APPROVED** [review]
great idea, but I share the repetition concerns. Approved if @djc 's suggestion is implemented.

**Comment by est31** [comment]
Also, please squash after the fixes. I see a merge commit in the PR :).

**Comment by uglyoldbob** [comment]
Looks like making a second version of the test_certificate_from_csr test or modifying that test to include the extended key usage would work. Any preference as to which one?

**Comment by cpu** [comment]
> modifying that test to include the extended key usage would work. Any preference as to which one?

That sounds reasonable to me :+1:

**Comment by uglyoldbob** [comment]
Support for the extended key usage probably needs to be added to CertificateParams::serialize_request.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
@uglyoldbob thanks, this is looking good!

@cpu, want to have a look before merging this? Already has @est31 approval.

(Let's bypass the merge queue to squash?)

**Comment by cpu** [comment]
> @cpu, want to have a look before merging this?

:+1: I will take a look today

File: rcgen/src/csr.rs
Comment (cpu) [inline rcgen/src/csr.rs:151]: I think this will silently drop instances of `x509_parser::ExtendedKeyUsage.other` holding `Vec<Oid<'a>>`.

Since there's no equivalent for an arbitrary EKU in rcgen I think it would be prudent to return `Err(Error::UnsupportedExtension)` if `!eku.other.is_empty()`

File: rcgen/tests/webpki.rs
Comment (cpu) [inline rcgen/tests/webpki.rs:501]: I'm a bit confused by this test. We put EKUs into parameters, check the expected EKUs are in the parameters, create a self-signed CA certificate from the parameters, and then verify the CA certificate's parameters contain the EKUs.

It doesn't seem like anything is testing that a CSR that has EKUs is loaded with those EKUs, or that a CSR created from params with EKUs preserve the EKUs. I think those are the important properties we want to test right?

Comment (uglyoldbob) [reply rcgen/tests/webpki.rs:501]: I was a bit stumped as how to fully write a test here since some of the testing I had thought of required for  CertificateParams::serialize_request to implement the extended key usage, but thought adding that might be better for a different pull request.

File: rcgen/src/csr.rs
Comment (uglyoldbob) [reply rcgen/src/csr.rs:151]: Oh that makes sense

File: rcgen/tests/webpki.rs
Comment (cpu) [reply rcgen/tests/webpki.rs:501]: > adding that might be better for a different pull request.

I would be OK with it being part of this PR!

File: rcgen/Cargo.toml
Comment (djc) [inline rcgen/Cargo.toml:39]: I don't think this what we want? It looks like this might be only for (some of) the tests? If so, we can make x509-parser a required dev-dependency.

File: rcgen/src/certificate.rs
Comment (djc) [inline rcgen/src/certificate.rs:817]: Please move this change into the earlier commit.

File: rcgen/Cargo.toml
Comment (uglyoldbob) [reply rcgen/Cargo.toml:39]: Oh I think I accidentally commited that part. It was so my ide would actually be helpful for the section I was working on.

Comment (djc) [reply rcgen/Cargo.toml:39]: I suggest running your IDE with --all-features (that's what I do).

**Review by cpu — APPROVED** [review]
Thanks! Had a couple small nits to mention, but I don't think they're blockers if you feel strongly.

**Review by cpu — APPROVED** [review]
Thanks! Would you mind squashing the two commits into one? It would save me having to admin merge this to do the squashing myself.

**Pull request merged**
